### DISCUSSION TITLE
tab zones

### DISCUSSION THREAD

**Discussion by N3C2L** [madprops/grasshopper#62] [Ideas]
If you drag and drop a tab onto another tab they build a "zone". When these two tabs are in a zone, a common colored edge (border) could surround them to make the zone visible. If you then drag and drop another tab in this zone the colored border will be extended around the newly inserted tab. The zone could be named, the name of the zone could be visible on top of the zone. And the tab zone could be slightly separated from the non- zoned tabs for an even better distinction. And why the zones? Because you can make topic zones for your interests. The zones wouldn't need to be visible when using other filters. The zones could have their own filter: "zones".

**Comment by N3C2L** [comment]
Maybe instead of the idea above : a "insert separator" cmd. With this command a separator in color of the background can be placed above or below a tab. Then between these separators you have a zone where you can move your tabs inside.

**Comment by N3C2L** [comment]
Ok, I found a workaround. Take a tab, delete the text. Change the favicon to the background of grasshopper. And done is a workaround separator. Of course new users would not know about this workaround, and would find a tab zone a better suit.

**Comment by madprops** [comment]
I personally dislike when extensions do this. When I tried moving items around in other tab managers, I didn't like how you could drop tabs in tabs instead of re-order. It makes re-ordering harder, and I personally don't have a need for zones.

However I think I'll add something....

**Comment by N3C2L** [comment]
Thank you very much. I look forward to trying out your addition.

**Comment by N3C2L** [comment]
Thanks for the split feature. I love how you can move the splits (by dragging the upper or lower tab). And cool that you can move tabs inside and outside the splits. I don't know if it's purpose but if you insert an upper and lower split to a single tab, then you can't move tabs inside this single tab. 
Another observation: If you select more tabs at once then I would execute another command that I miss: "Create one split on top and one split on bottom of the selected tabs". In the actual version if you select more than one tab and run "split top", it creates splits on every tab selected, but I want only 1 split on top of the selected tab  and 1 split on bottom of these selected tabs (for example: selection of 10 tabs: the first tab on top will get a split and the last tab gets a split on bottom, but not the second tab or third tab and so on would get any splits). Only one split on top of these selected tabs should be created. And only one split on bottom of these selected tabs should be created. Then I would have for example 10 tabs with one split on top and one split on bottom so I can move tabs inside and outside this 10 tabs.

May I suggest some additional things regarding this cool split feature:

1. Option to change the split line color
2. Right click on a tab > remove single split from the selected tab

**Comment by madprops** [comment]
I took notes. But I don't understand the last point (2). Since there's already a right click menu for that, to remove the split, it's in More in the item menu.

**Comment by N3C2L** [comment]
Thanks. Sorry, I forgot to mention: If you have created more than one split, then execute the remove the split cmd, the prompt asks you to delete every split, not a single one.

**Reply by madprops** [reply]
Oh I see what you mean. I was doing the command wrong.

**Reply by N3C2L** [reply]
Thanks for fixing. What do you think about the option for adding names to a top split? Then I could faster recognize the content of the tabs below the top split. For example I would name a top split "books" then I would know the tabs below are about books. The name could maybe positioned above the split line, left-justified and with slightly more margin above the name to give more space.

**Reply by madprops** [reply]
Hmm, not a bad idea, I'll see if it works.

**Reply by N3C2L** [reply]
Thanks!

**Comment by madprops** [comment]
Ok lots of modifications and fixes on Splits. New settings in the `Edits` category at the bottom.

**Comment by N3C2L** [comment]
Thanks for the new splits settings in the "Edit" Category. The new "split auto" cmd was exactly what I was looking for, thanks. 
Small issue I observed: If only one tab is selected and then the "split auto" cmd is executed, no split will be created. Maybe a popup could inform the user, that the split auto cmd only works if more tabs at once are selected.

**Reply by madprops** [reply]
Ok. I'm working on big updates  on split. Upload soon.

**Reply by N3C2L** [reply]
Perfect, thanks!

**Comment by madprops** [comment]
Ok added and fixed some stuff, including the addition of Split Titles which you can add to the top split.
The titles can be displayed at the top always, or on hover.

**Comment by N3C2L** [comment]
Big thanks for the naming feature. Having it visible in the middle of the tab is good, because then it doesn't overlap with the favicon.

**Reply by madprops** [reply]
Next version will allow moving the text to the left, center, or right.

**Reply by N3C2L** [reply]
Very useful, thanks.

**Comment by N3C2L** [comment]
I think the naming feature of a split could be improved, buy implementing a new
split tab: If a top split (or split auto) will be created, a new internal tab will be created above the first selected tab. This new split tab has an internal URL (starting with "moz-extension://..").
The advantages with a new split tab could be: 
1. The split name is not linked to an existing tab with a real website anymore
2. Better filtering (maybe with an new "split" filter category, showing only the internal split tabs. When filtered "split" the split tabs could show a new button on themselves with the function to leave the "split" filtering and navigating to the unfiltered split view (as shown in the all tabs filter view)
3. Auto split creation could work when only one tab is selected (because the internal tab will be created additionally to the single tab)
4. The visible website content of this internal new split tab could be the given name of the split tab
5. The name of the split could be renamed by editing the split tab itself in the tab list

There are extensions which create a new internal tab for example searching bookmarks:
https://github.com/aaFn/Bookmark-search-plus-2

**Comment by madprops** [comment]
Don't like the idea of a dummy tab too much, bottom split tabs would also need a dummy tab too. What I did was add Replace Split commands which you can use to give the split properties from one tab to another, in case you need to remove or move a top/bottom split tab. I can enhance these replace functions more like dragging with ctrl+shift pressed or similar.

Also it's easy to make special separator tabs using Domain Rules, like giving separator.com/MyTitle a special color etc, and use instances of that as separators.

But idk, it does make some sense, I'll think about it, maybe as an addition.

**Comment by N3C2L** [comment]
Alright, thanks for the feedback. I would really appreciate it if you can add it.
With your additions I could rename and filter splits (see points 2 and 5 above).

**Comment by madprops** [comment]
What I'm going to do is add special blank tabs.  Every about:blank tab will be prepared in a special way.

1) They get a random initial title that is {consonant} {vowel} {consonant} {vowel} 

2) They get a random color

3) They get to have a slightly bigger font size

4) They don't have an icon/favicon

5) They get a dashed outline to set it apart further

6) They get a slight background color difference

This way you can simply use blank tabs as separators or however you want.

Of course you can edit the title or color after it's created.

**Comment by N3C2L** [comment]
Thanks for the blank tabs. They are useful. I think a new cmd would then be handy with this blank tabs:
Maybe a cmd which contains these steps:
(for example if you have 5 tabs selected before)

1. Create the blank tab above the tab selection (the selected 5 tabs)
2. Include this blank tab in the tab selection
3. run the split auto cmd
4. spawn the rename prompt of the blank tab

**Reply by madprops** [reply]
There's gonna be a insert blank tab above tab, command. I don't want to mix blank tabs with splits.

**Reply by N3C2L** [reply]
Ok, and what about spawning the edit title popup for the first tab (no blank tab) below the split (after the split auto cmd was executed).

**Reply by madprops** [reply]
What I'm thinking is removing split titles and just use blank tabs for that...

**Reply by madprops** [reply]
But it would be same problem with the splits being attached to a tab. I'll keep it as it is.

**Reply by N3C2L** [reply]
But I would love the option that splits would be attached to blank tabs, because they are more flexible than the split titles.

**Reply by madprops** [reply]
Question is, why have splits then? Just create blank tabs to categorize your tabs?

**Reply by madprops** [reply]
I guess you want to associate blanks with splits for filtering

**Reply by madprops** [reply]
It could then be a zone between a blank and a split bottom....

**Reply by N3C2L** [reply]
I would rename the blank tabs of course same as the split titles. But additionally they could be filtered. Maybe with a new filter category like when coloring tabs but for the blank tabs.

**Reply by N3C2L** [reply]
Yes you are right a zone between a blank and a split bottom

**Reply by madprops** [reply]
I'm making it so clicking the blank tab, highlights all items between it and a split bottom

**Comment by N3C2L** [comment]
They could be filtered.

**Reply by madprops** [reply]
How do you see the filtering can be done?

**Reply by madprops** [reply]
What happens if I filter it and click it?

**Reply by N3C2L** [reply]
Only the blanks will be visible (like when "filter red" cmd then only red colored tabs are visible).

**Comment by madprops** [comment]
I'm making it so Auto split creates a blank at the top, and inserts a bottom split on the last selected item. But it's gonna be auto, you can edit title later. It uses 2 random words automatically.

**Reply by madprops** [reply]
Split auto only deals with splits now. Insert Blank will act as you want when selecting multiple items. I'll upload soon.

Making it so when filtered it shows all tabs and selects the blank. If not filtered, it selects all related items.

**Reply by N3C2L** [reply]
Cool thanks, looking forward to try it out.

**Comment by madprops** [comment]
New version `v2121` is up. I like how it turned up. Thank you for the ideas.

**Reply by madprops** [reply]
Fixes coming soon.

**Comment by N3C2L** [comment]
Thank you very much. I couldn't test it too long because It's getting late here and I will going to bed now. But I tested it and like it so far, also the filtering is very nice.

**Comment by N3C2L** [comment]
Thanks for the great header features. It's a much better name than "blanks".

To reduce clicks even more, may I suggest some additional things:

1. If I select more than one tab and will then run the header cmd, it should additionally spawn the edit title prompt for this header automatically. It could be a setting.

3. If I select more than one tab and will then run the header cmd, it creates the split line only on bottom but not above the header. I would really like it if you could add this line (border) above the header. Maybe an additional option in the settings to add this border and the thickness for the header.

**Reply by madprops** [reply]
1) I can do that

2) Yes right now it treats the header as the top border/split. I don't want to give it a normal split top because i don't want to auto save edits to the tab since it can be used later as a normal tab with another url. What I can do is alter its style for headers to something better, what do you have in mind for the style?

3) Same as 1.

**Reply by N3C2L** [reply]
Thanks, I edited my post above and deleted the first one, but if you can implement it I would be glad.
About the styling: Maybe the same styling options as for the split line: border, padding, border width.

**Reply by madprops** [reply]
New change. I no longer use `about:newtab` Now it uses a `moz-extension` page. So new tabs no longer become a header.

I'm not sure how I'll do the title thing, it's a bit tricky. I'll continue tomorrow.

**Reply by N3C2L** [reply]
Alright, thanks.

**Comment by N3C2L** [comment]
Thanks that you implemented it! This is a pleasant addition and it works wonderfully.

What do you think about the idea adding a header directory? I was thinking about the tab box as an inventory for this. In the tab box there is the "recent tabs" button on top. Maye to the right of this button could be another button called "headers". When clicked, the content of the tab box would switch showing only the header tabs there. Maybe the header tabs need other favicons but I don't know which would fit.

**Comment by madprops** [comment]
You can now click the Tab Box title and select Headers.

Also I changed the setting categories. There's no more `Edits`.

Settings moved to `Zones`, `Colors`, and some to `General` and `More`.

**Comment by N3C2L** [comment]
Many thanks. You are really a great help and are doing me and the other users a lot of good with your extension.

Small Issues I observed: 

1. The favicon of an inserted header could be the traffic light symbol (I mean the same symbol from the palette for headers). It looks better than the generated one.
2. If the favicon of a header is customized with your cool new cmd "edit icon", than it is normally visible in the tablist, but not visible in the tab box.

I'll continue tomorrow commenting if I find something, but thanks again.

**Comment by N3C2L** [comment]
I would need an option for a specific header filtering where the header tab and only the first tab under this header are visible. Because only then would I know which tab would be loaded if I clicked on a header. Perhaps a header filtering setting that is not the default but can be enabled.

Because currently, only the header is filtered, but not the tab below. However, I would like to see both together in the filtering so that I have a better overview. This way, I would see all headers and only their respective first tab underneath to have an overview of where I want to navigate.

I don't know how the header + first tab can be filtered together. But perhaps these tabs can be internally numbered. So, describe the tabs with numbers as they are arranged in the tab list. For example, the 12th tab from the tab list has the number 12, and the first tab under it has the number 13; these two are displayed in the header filtering. If another header has the number 4, and the tab underneath has the number 5, these should also be displayed in the filtering.

I would like to have the option for auto-restore set to "never" for the tab under the header. But the header remains auto-restore as it is now. This way, both options are available in this filtering, which I would need so that I do not accidentally exit the filtering, but also have the ability to leave the filtering by clicking the header.

**Reply by madprops** [reply]
I added context for the Header filter. You can adjust this in Filter settings. I made the default 2.

**Reply by N3C2L** [reply]
This is great, thanks! That you can customize how many context level you want is awesome!

Just an observation:
If I am in the header filtering and click a header there I will navigate to all tabs view which is good. But is there a cmd which brings me back to the previous filtering? Because step back is not bringing me back to the header filtering and previous mode / next mode also doesn't navigate me back to the previous filter. Do you think a previous filter cmd / next filter cmd would be good?

**Reply by madprops** [reply]
>Do you think a previous filter cmd / next filter cmd would be good?

I'm not sure ... would it be only the filter mode like `Header` or would that include the text typed in the filter too?

**Reply by N3C2L** [reply]
Maybe two additional separate cmds would make sense but I don't know how to name them.

**Reply by N3C2L** [reply]
But personally I would include filter modes like headers and text typed filter in one cmd (prev / next filter)

**Reply by madprops** [reply]
I could save the filter used before going back to `All` and go back to it. I don't know what "next filter" would be.

**Reply by N3C2L** [reply]
Ok, that would work. But I would prefer a separated thing more. A common history of the cmds like filter (headers etc.) and the typed filter. And the previous and next filter cmds would then only switch trough this common history (from the start of the history and the end what was filtered last)

**Reply by madprops** [reply]
That doesn't sound too simple. I'm adding a Previous Filter command for now, to return 1 step.

**Reply by N3C2L** [reply]
Alright, thanks! Yes, return 1 step is most important, because then I can switch back and forth (like alt+tab) between all tabs and header tabs.

**Comment by N3C2L** [comment]
Hmm. Just know I need two cmds to simulate that alt+tab thing. It's the cmd "previous filter" and "show tabs". I would be glad if you could combine these 2 cmds into a new cmd maybe "filter switcher". This would act as an alt+tab, so when you are in the all tabs view it would switch to previous filter, and when you are in an active filtering view it would you take you back to all tabs. The ESC key can't do this now.

**Reply by madprops** [reply]
Ok I think I can make Previous Filter toggable, i'll try it.

**Reply by N3C2L** [reply]
Thank you very much, I'm looking forward to your changes.

**Reply by madprops** [reply]
It should work better now. Also added a setting `clear_on_all`, true by default, when clicking `All` it clears the text filter too. Also made `Previous Filter` the middle (aux) click command of the `Filter Menu`. `Previous Filter` should toggle better now.

**Reply by N3C2L** [reply]
Excellent changes, thanks. I think I map the previous filter cmd to double shift.

**Comment by N3C2L** [comment]
May I suggest this enhancement for the header tabs:
I like the tooltips over hovered tabs showing the tab title & URL. Of course URLs would not make sense in header tooltips.
But my longer header titles are sometimes cropped in the grasshopper interface. I would be glad about this addition: If you hover over a header tab, show his full title in a tooltip.

**Reply by madprops** [reply]
Ok

**Reply by N3C2L** [reply]
Thank you!

**Comment by N3C2L** [comment]
Regarding the filter "loaded":
It would make sense to exclude the header tabs from there. Because it's confusing clicking on a header tab there, and then a tab will load which isn't loaded.
Maybe the header filter could be separated in headers loaded (if header and the tabs below are loaded) and headers unloaded (tabs where the header and the tabs below are unloaded). For a better overview.

**Reply by madprops** [reply]
I'll remove headers from loaded

**Reply by N3C2L** [reply]
Thank you!

**Comment by N3C2L** [comment]
I really like the split line above the header tab in the All view. Also, I find it great that the title of the header is centered in the All view. However, I have an issue with the overview during filtering. Let's say my header and the tab below it are given a blue color. Then I navigate to the Filter Blue view: And here, the header tabs are narrow and too similar to the other tabs. I would greatly appreciate an option where the header tabs when filtered appears with the same size, split line, spacing, and centered text as in the All view. Otherwise, the header is displayed in the filter just like any other tab and does not stand out from the others, which is as important as in the All view for better clarity.

There is another idea in my mind regarding the background: Custom Background color if a filtering view has changed from "All" to any other filtering view (tags, colors,..). This filter background color could may be adjusted and enabled below the background color in the "Theme" settings. That could be useful for quickly recognize if I am in a filtering view or not. "All" view could use the normal background color, and any other filtering view the filter background color (which overrides when enabled the normal background color).

**Comment by madprops** [comment]
Headers stand a bit more now since they have the centered text when filtered, and have the top border.

Added `Filter Effect` setting in `Filters`. Default : true
When enabled it will show inverted colors in the filter input, when filtered.

**Reply by N3C2L** [reply]
Thanks a bunch!
The filter inverted solution is appreciated.
And thanks for the alterations on the headers when in a filtering view.

**Reply by N3C2L** [reply]
Just a small issue:
The border on top of the filtered header is a little hidden by the favorite bar. You can see this better when split width is set to 6, and now compare the all view and the filtered view.

**Reply by madprops** [reply]
Not sure what you mean, it looks normal to me, what else can I change to see this?

![gh3425](https://github.com/madprops/grasshopper/assets/206498/981f7831-9669-438c-ba90-552c0ce925c1)

**Reply by madprops** [reply]
Ok maybe it's eating 1 pixel.. i'll take a look.

Fixed in next version.

**Reply by N3C2L** [reply]
Yes, it was just a small area there.
Thanks, it's better now.

**Comment by N3C2L** [comment]
I find the Header Tabs really useful, thank you.
However, I think there could be an even finer visual breakdown. I mean that there could be a second subordinate hierarchical level to organize tabs into subtopics.

Therefore, I thought about Subheaders (you can name them differently if you like). These could be like the Headers but smaller and less prominent. They would be useful to help create sub-chapters under the Headers. In these sub-chapters, you can further categorize your tabs.

I have thought of these attributes:

1. The size should be the same as regular tabs, without spacing like the header tab.
2. The same background color as the header tabs.
3. No split line.
4. Downward arrow as the favicon (why the arrow? The downward arrow is meant to point to the tabs below).
5. The title left-aligned to make it less likely to be confused with the normal header tab.
6. When inserting a subheader, the rename prompt could appear, similar to the header tab.

I would be really glad if you could implement this.

**Reply by N3C2L** [reply]
I really don't want to bother you here, but my concern above is important because my split line from the header is thick (setting 6) and the spacing is there too. And now, imagine if such headers were always present between smaller headings, then that would be too much. That's why I suggested the Subheaders. I promise not to post any further concerns about the tab zones in the next few days.

**Comment by madprops** [comment]
Introducing another special tab is not so simple. What I'm going to do is expand Header a bit, add some options to configure it, like with or without a top border etc. I'm thinking some prompt when creating a header like "Add this thing? yes/no"

**Reply by N3C2L** [reply]
Yes, that would be really great, thank you. But can't the same header with different appearances be loaded on two different cmds (first cmd for header 1 with appearance 1, second cmd for header 1 with appearance 2) ? That way, the prompts would be eliminated. Perhaps these things could then be adjusted in the header settings?

**Reply by madprops** [reply]
You do make a good point there

**Reply by madprops** [reply]
You can try the Subheaders now

**Reply by N3C2L** [reply]
Thank you for the new Subheader!

So that I wouldn't have to bother you for small CSS things, I would be grateful if you could provide a new class for the Subheader.

I think I can't change that favicon with CSS, that's why I am asking:

Thanks for the favicon with the arrow down in the Subheader. However, the blue color in it is a bit straining on the eyes. Would it be possible to use the arrow from the download link below instead? That would help me.
https://www.flaticon.com/free-icon/download_5612000?term=arrow+down&page=1&position=14&origin=search&related_id=5612000

**Reply by madprops** [reply]
You can target headers in this way:

1) Full headers have these classes `header_item` and `split_top`. 
So you can target it like `.header_item.split_top`

2) Subheaders simply don't have the `split_top` class.
So you can do something like ".header_item:not(.split_top)"

I don't want to add icon images, I don't do that at all right now.

What I'm doing is changing it to an svg triangle.

**Reply by madprops** [reply]
Also moved the zone item context menu items to the main context menu under the `Zones` category. This is up now.

**Reply by madprops** [reply]
I'm gonna remove the svg. What i'll do is just add icon settings for header and subheader.

I'll make header/subheaders fallback to svg arrow.

**Comment by N3C2L** [comment]
Thank you for the CSS class explanations! I will try them out.

The solutions with the icons are excellent. This way, everyone can choose whether they want the triangle or other emojis.

A small observation: If you have only one tab selected and then execute the "Split Both" command, nothing happens. It would be helpful at this point if a popup appeared indicating that at least two tabs need to be selected for this action. This could assist new users who might otherwise be confused.

One more thing is important to me: I would like to be able to select headers without them immediately loading the tab underneath, providing an option for users like me. Why would this be helpful to me:
1. The "Autohide Taglist" function could be used for headers as well (because if you click on the header, you wouldn't see the tags inside otherwise).
2. With the headers in the tab box, I could click on them to get an overview without the first tab loading immediately.
3. If I select a header in the left Grasshopper interface of Floorp (filtering "header"), it wouldn't immediately restore there, but would be displayed well in the right Grasshopper interface (filtering "all").

**Reply by madprops** [reply]
Alright I think I'll just change the behavior and just treat headers as normal selectable tabs.

**Reply by N3C2L** [reply]
Wonderful, thanks.

**Reply by madprops** [reply]
Do you think it's better if clicking the header shows a custom html page, or should it avoid focusing the tab and just select it?

**Reply by N3C2L** [reply]
I'm not entirely sure, but I know that when the custom HTML page is displayed, it is correctly selected in the right Grasshopper in Floorp.

**Reply by madprops** [reply]
I'll make it only select, I think it's better. Or well just add a setting

**Reply by N3C2L** [reply]
Yes, an additional setting would be great!

**Reply by madprops** [reply]
Alright new version is up. Added a couple of settings in Zones at the bottom.

**Reply by N3C2L** [reply]
Excellent, thank you very much. Everything works great.

**Comment by N3C2L** [comment]
Thank you for the "Header Action" settings in the zones.

Maybe a bug:
Regarding this 2 settings in the color settings category: 
Active tabs (text) + Active tabs (background)
If this 2 settings above are checked, they should also affect a header when clicked. (At least when the header action is on "activated") 

A note on the header custom html page:
I find it practical that I can change the background color depending on my mood. 
But sometimes I accidentally click the buttons in the content of the website, and then the colors change again, and I have to adjust them again. Would it be possible to add a lock button there, maybe in a corner? Then I can set everything the way I want, press the lock button, and the colors will stay the same even if I accidentally click on the buttons.

**Reply by madprops** [reply]
Added a setting in colors to enable coloring of headers. I think headers shouldn't be altered by default, but if you enable the checkbox colors will apply to them as well.

Improved the header page. Added some toggles on the top right.

**Reply by N3C2L** [reply]
Thanks a lot for these great implementations.

**Comment by madprops** [comment]
Removing "color header" but adding color settings for header and subheaders. Also making the selected effect on headers "underline", so it has no clashes with colors.

**Reply by N3C2L** [reply]
Thanks, but I would be glad if you could add the active tab color for the header as well (There was a checkbox in the color settings before to enable them).

Edit: Nevermind, the header get already the custom active tab color (if set).

**Comment by N3C2L** [comment]
This idea has been implemented, so I can close this thread.
